Add updateSealedTx and ExtraTxBodyContent

Anviking opened the pull request:
- [x] Refactoring: Replace `to{Shelley, Allegra, Mary, Alonzo}TxOut` with `toCardanoTxOut` for easier re-use.
- [x] Add `updateSealedTx :: SealedTx -> ExtraTxBodyContent -> Either ErrUpdateSealedTx SealedTx`
- [x] Test that `updateSealedTx noExtraBodyContent` works with all PAB goldens
- [x] Test that `updateSealedTx noExtraBodyContent` fails when there are existing key witnesses
- [x] Test that inputs and outputs are combined correctly. 

### Comments

Not tested:
- That it works for all previous eras.

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1140

Relates to / should be needed by #2906

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking committed:
Implement updateSealedTx and ExtraTxBodyContent

- [x] Add `updateSealedTx :: SealedTx -> ExtraTxBodyContent -> Either ErrUpdateSealedTx SealedTx
- [x] Test that `updateSealedTx noExtraBodyContent` works with all PAB goldens
- [x] Test that `updateSealedTx noExtraBodyContent` fails when there are existing key witnesses
- [x] Add property test, showing that the resulting inputs and outputs
    are the union of the inputs and outputs of the arguments.

Not tested:
- That it works for all previous eras.

Anviking committed:
Replace per-era helper with one toCardanoTxOut

Pure refactoring. This allows the case-switch-on-era logic to be
re-used.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3472,8 +3472,14 @@ instance IsServerError ErrBalanceTx where
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "The transaction could not be balanced, because it contains "
+                , T.pack (show n), " "
+                , "existing key-witnesses which would be invalid after " -- TODO: I /think/ key-witnesses is an accurate term here, but haven't checked.
```
In the ledger it seems the witnesses are all equal:
https://github.com/input-output-hk/cardano-ledger-specs/blob/0eb9b38bb125ec8eedf0f7aabf9e09f9b7bf79f7/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs#L181-L188

But it seems we only get key witnesses from cardano-api:
https://github.com/input-output-hk/cardano-node/blob/0ed04ab8d0a205691219f04623a3d880545a3f34/cardano-api/src/Cardano/Api/Tx.hs#L493-L507

This might actually be a separate problem: any existing non-key witnesses will be lost by the balancing…

Anviking replied:
It turned out that script witnesses, datums, etc were stored in the cardano-api version of the `TxBody`. So all is good.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -1600,3 +1617,111 @@ matrixNormalTxExamples =
+                        , "future-pay-out-5.json"
+                        , "future-pay-out-6.json"
+                        , "future-settle-early-2.json"
+                        , "future-settle-early-5.json"
+                        , "future-settle-early-6.json"
+                        , "game-sm-success-2.json"
+                        , "game-sm-success-4.json"
+                        , "game-sm-success_2-2.json"
+                        , "game-sm-success_2-4.json"
+                        , "game-sm-success_2-6.json"
+                        , "multisig-failure-2.json"
+                        , "multisig-sm-10.json"
+                        , "multisig-sm-11.json"
+                        , "multisig-sm-2.json"
+                        , "multisig-sm-3.json"
+                        , "multisig-sm-4.json"
+                        , "multisig-sm-5.json"
+                        , "multisig-sm-6.json"
+                        , "multisig-sm-7.json"
+                        , "multisig-sm-8.json"
+                        , "multisig-sm-9.json"
+                        , "multisig-success-2.json"
+                        , "ping-pong-2.json"
+                        , "ping-pong-3.json"
+                        , "ping-pong_2-2.json"
+                        --, "prism-3.json" -- Error in $[0]: there should be one 'lovelace' in 'value'
+                        , "pubkey-2.json"
+                        --, "stablecoin_1-2.json" -- Error in $[0]: Value should not be empty
+                        , "stablecoin_1-3.json"
+                        , "stablecoin_1-4.json"
+                        --, "stablecoin_2-2.json" --Error in $[0]: Value should not be empty
+                        , "stablecoin_2-3.json"
+                        , "token-account-2.json"
+                        , "token-account-5.json"
+                        , "uniswap-10.json"
+                        , "uniswap-2.json"
+                        , "uniswap-7.json"
+                        --, "uniswap-9.json" -- Error in $[0]: there should be one 'lovelace' in 'value'
+                        , "vesting-2.json"
+                        ]
```
Duplicated with CompatibilitySpec… Could de-duplicate via some shared `TestData` or `Util` module if we want to

paweljakubas replied:
I know. In the first version I wanted to dump the plutus script. Definitely something could be done to reuse those jsons name as they repeat in 3 places already. Feel free to de-duplicate if you have any idea and is eager to to do it withing this PR

Anviking replied:
Wasn't super-eager, but for the sake of making the property tests neater, I might give it a shot in a follow-up PR.

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3478,8 +3478,16 @@ instance IsServerError ErrBalanceTx where
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "The transaction could not be balanced, because it contains "
+                , T.pack (show n), " "
+                , "existing key-witnesses which would be invalid after "
```
I would even say that at this stage we should not have any witnesses collected. signing should be done only after tx is balanced (hence I slightly incline to introduce AlreadyBalancedTx error in balance tx PR)

Anviking replied:
I added a `Please sign the transaction after it is balanced instead.` line afterwards.

However, if for some reason a user want to add key-witness, I'm not sure if the balance-endpoint can take the size of the witness into account for the coin selection (if it is not known to the wallet). Maybe it does…

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1251,70 +1248,77 @@ toCardanoLovelace (W.Coin c) = Cardano.Lovelace $ safeCast c
-        , Cardano.AddressInEra Cardano.ByronAddressInAnyEra
-            <$> deserialiseFromRawBytes AsByronAddress addr
-        ]
+    toShelleyTxOut :: HasCallStack => W.TxOut -> Cardano.TxOut ShelleyEra
```
I wonder if we could not reuse here better the functions `toEraTxOut`. And apply the trick was applied here in `toGenTx` https://github.com/input-output-hk/cardano-wallet/pull/2874/files

Anviking replied:
Not sure what you're suggesting here. `toCardanoTxOut` is basically a `toEraTxOut`, no?

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -441,6 +453,142 @@ mkDelegationCertificates da accXPub =
+            -> Ledger.TxBody (Cardano.ShelleyLedgerEra era)
+        adjust (ExtraTxBodyContent extraInputs extraOutputs modifyFee) era body = case era of
+            ShelleyBasedEraAlonzo -> body
+                    { Alonzo.outputs = Alonzo.outputs body
```
no adding extra outputs here? it seems wrong at first glance

paweljakubas reviewed (approved):
good work! We can probably simplify error and say there are witnesses without specifying the number. Also please double check the possible missing output update for alonzo - I think suggested below property test in point 2 would catch that!

Also suggestion for another property tests:
1. we have extra input(s) and update -> after updating we always find after decostruction those inputs
2. the same for extra outputs
3. the same when we update fee - the new value will be there
4. also combinations of 1-3
5. maybe check additionally after point 4 that binary payload (ie. BS.length) is always bigger after updating  
And do those for different eras. Maybe we could reuse generators from cardano-api - there are some exposed for txOut and txIn I believe (if possible)

Anviking committed:
Mark prop_updateSealedTx pending

Changing the dummy address length fixes it, but causes a cascading set
of other golden-changes which are serious enough to get their own PR.
